### ISSUE TITLE
Proposal: Automatically self destruct after 24 hours when alpha features are enabled

### ISSUE THREAD

**Issue by negz** [crossplane/crossplane#5797]
### What problem are you facing?

https://docs.crossplane.io/latest/learn/feature-lifecycle/#alpha-features

Crossplane uses a feature lifecycle similar to upstream Kubernetes. When we first ship a feature, we mark it alpha. It's off by default, and we tell folks it's not recommended for production use.

We ship features as alpha to test them and learn whether:

* They appear to solve a real problem for folks
* We got the API and concepts right
* They're relatively stable and performant

Our contract around alpha features is that we can make breaking changes or even remove them at any time. The goal behind this contract is to lower the bar to get features in peoples' hands. To lower the bar to let folks experiment and give us real feedback. We never ship an alpha feature we _already know_ isn't what we want, but knowing we don't have to support [^1] alpha implementations allows us to move a little faster and spend less time agonizing over the design phase. It removes the expectation that we'll get the feature right on the first go.

Once (and indeed _if_) an alpha feature reaches beta, we make a stronger commitment to avoid breaking folks.

Unfortunately, I believe the alpha lifecycle stage is hurting the project more than it helps it. I see a couple of related problems, that are all symptoms of the fact that folks don't wait until a feature is in beta to use it in production:

* Folks become dependent on how a feature works. Removing it or making breaking changes is disruptive, which upsets them and paints the project in a negative light.
* There's no pressure on feature contributors to prioritize maturing (or changing, or removing) a feature once it ships as alpha.

These problems compound. The more folks use the alpha features, the less pressure there is to mature them, the longer they exist at alpha, the more folks use them, etc. Ultimately this leads to situations where we can't make meaningful changes to alpha features without either disrupting a lot of folks, or spending time and effort to avoid disrupting them. Essentially, we end up having to treat alpha features as if they weren't alpha, thus defeating the point of having them in the first place.

### How could Crossplane help solve your problem?

I think alpha features are a good idea in theory, but we need a way to remind users and contributors that shipping an alpha doesn't mean the feature is done. It's like shipping a design document - there's more work to do before folks should actually use it!

One way we could potentially enforce this way of thinking would be to make it harder to use alpha features for anything other than experiments and feedback. For example configure Crossplane to automatically shut down (and remove all of its resources?) after 24 hours. This is similar to [GKE alpha clusters](https://cloud.google.com/kubernetes-engine/docs/concepts/alpha-clusters), which self destruct after 30 days.

[^1]: Support in the sense of maintaining backward API compatibility, behavior compatibility, or even guaranteeing the feature will keep existing in anything like its alpha form.

**Comment by mproffitt** [comment]
I would posit that this issue still has a lot of relevance although I dislike the proposed solution (and will endeavor to explain why).

Alpha features are a great way to introduce features in general and get early feedback but I see a very different issue to them being made available in that it appears that you often push alpha features onto users, presumably to get that early feedback, but without making it explicitly clear "look, this is in alpha, we think it will solve your problem, try it out and let us know if it helps or not". 

As a case in point on this, take the alpha feature `Usage`. Just a quick search of slack will reveal at least a couple of threads where the answer is simply "You can use `Usage`" without any elaboration on "This is an alpha feature". Whilst it is true, that statement is covered in the documentation, the simple statement of "You can do this" sets the expectation that the feature works and is stable enough to use.

The idea of having self-destructive crossplane is, IMO, dangerous without some way to say "I know the risks here".

I will often run crossplane in `kind` maybe for a few days or even weeks as I'm experimenting with ideas / building compositions and designs and the idea of having crossplane completely self destruct potentially when I'm in the middle of trying out an idea would build a great deal of frustration towards a product that otherwise does exactly what I want from it, especially if I'm doing anything complex as it takes time for infrastructure to build. (Example, building a new composition to provision inside existing infrastructure - if crossplane automatically wiped out infra that takes 20m to build, that's 20m I'm doing nothing towards my project.

Whilst I'm averse to running alpha features in any environment, as an alpha feature moves closer to its beta release, I will start developing against that. Again the example here is `Usage` - this is earmarked for v1.18 so I will start looking at where and how I integrate them to existing compositions as we move closer to that release.

TL;DR I think a large part of the issue with Alpha features is down to how they are communicated to the community.

**Comment by negz** [comment]
/fresh

> TL;DR I think a large part of the issue with Alpha features is down to how they are communicated to the community.

Let me know if you have any ideas of how to communicate it better.
